File: scripts/common/app-config.ts
import fs from "node:fs";
import { readJson } from "./io.js";
import { rootPath } from "./paths.js";

export type DropCluster = "devnet" | "mainnet-beta";

export type AppConfigV2 = {
  cluster: DropCluster;
  clusterFlavor: DropCluster;
  backendBaseUrl: string;
  identityUri: string;
  iconUri: string;
  siwsDomain: string;
  siwsStatement: string;
  privacyPolicyUrl: string;
  supportUrl: string;
  termsOfUseUrl: string;
  collectionName: string;
  rpcPrimary: string;
  rpcFallback: string;
  candyMachine: string;
  candyGuard: string;
  collectionMint: string;
  collectionUpdateAuthority: string;
  skrMint: string;
  proceedsWallet: string;
  proceedsSkrAta: string;
  mintPriceSkrBaseUnits: string;
  allowlistStartIso: string;
  publicStartIso: string;
  allowlistPerWallet: number;
  publicPerWallet: number;
  botTaxLamports: number;
  merkleRootBase58: string;
  allowlistWalletCount: number;
  policyAnnouncement: string;
};

export function getAppConfigPath(cluster: DropCluster): string {
  return rootPath("artifacts", `app-config.${cluster}.json`);
}

export function readAppConfig(cluster: DropCluster): AppConfigV2 {
  const path = getAppConfigPath(cluster);
  if (!fs.existsSync(path)) {
    throw new Error(`Missing app config artifact: ${path}`);
  }
  return readJson<AppConfigV2>(path);
}

File: android-native/app/src/main/java/com/seekerravens/mint/core/AppConfig.kt
package com.seekerravens.mint.core

import android.content.Context
import android.net.Uri
import com.google.gson.Gson
import java.net.URI

class AppConfigLoader(
    private val context: Context,
    private val gson: Gson
) {
    fun load(assetName: String = "drop-config.json"): AppConfig =
        context.assets.open(assetName).use { input ->
            gson.fromJson(input.reader(), AppConfig::class.java)
        }
}

data class AppConfig(
    val cluster: String,
    val clusterFlavor: String,
    val backendBaseUrl: String,
    val identityUri: String,
    val iconUri: String,
    val siwsDomain: String,
    val siwsStatement: String,
    val privacyPolicyUrl: String,
    val supportUrl: String,
    val termsOfUseUrl: String,
    val collectionName: String,
    val rpcPrimary: String,
    val rpcFallback: String,
    val candyMachine: String,
    val candyGuard: String,
    val collectionMint: String,
    val collectionUpdateAuthority: String,
    val skrMint: String,
    val proceedsWallet: String,
    val proceedsSkrAta: String,
    val mintPriceSkrBaseUnits: String,
    val allowlistStartIso: String,
    val publicStartIso: String,
    val allowlistPerWallet: Int,
    val publicPerWallet: Int,
    val botTaxLamports: Int,
    val merkleRootBase58: String,
    val allowlistWalletCount: Int,
    val policyAnnouncement: String
)

fun AppConfig.resolvedBackendBaseUrl(): String = ensureTrailingSlash(resolveUrl(identityUri, backendBaseUrl))

fun AppConfig.resolvedIdentityUri(): Uri = Uri.parse(resolveUrl(identityUri, identityUri))

fun AppConfig.resolvedIconUri(): Uri = Uri.parse(resolveUrl(identityUri, iconUri))

fun AppConfig.activeRpcUrl(): String {
    return listOf(rpcPrimary, rpcFallback)
        .firstOrNull { it.isNotBlank() && !it.contains("YOUR_KEY") }
        ?: "https://api.devnet.solana.com"
}

internal fun resolveUrl(base: String, value: String): String {
    if (value.startsWith("http://") || value.startsWith("https://")) {
        return value
    }

    return runCatching {
        URI(base).resolve(value).toString()
    }.getOrElse {
        value
    }
}

internal fun ensureTrailingSlash(value: String): String =
    if (value.endsWith('/')) value else "$value/"
